### ISSUE TITLE
matmul hangs / falls into 50-80x slowdown on XPU once reduction dim k crosses 2**24 (oneDNN JIT GEMM 'Invalid immediate value')

### ISSUE THREAD

**Issue by cakedev0** [intel/torch-xpu-ops#4805]
Hi, I'm working on scikit-learn and on benchmarking scikit-learn and some implementation variants (scikit-learn intelex and array API backed scikit-learn). I found this bug during some benchmarks.

### 🐛 Describe the bug

`torch.matmul`/`@` on XPU hangs (no error, no return) for a `(m, k) @ (k, 1)` matmul once `k` crosses `2**24` (16,777,216) for `float64` input - but only for some values of `m`. Verbose oneDNN logging shows the JIT GEMM kernel generator failing repeatedly with `Invalid immediate value` right before the hang, which looks like the actual cause: something derived from `k` (or a stride) doesn't fit in a 24-bit immediate field once `k > 2**24`, and instead of falling back to a different kernel it appears to retry the same failing one indefinitely.

```python
import torch

k = 16_777_217  # 2**24 + 1
m = 2
A = torch.randn(m, k, dtype=torch.float64, device="xpu")
B = torch.randn(k, 1, dtype=torch.float64, device="xpu")
C = A @ B  # hangs
torch.xpu.synchronize()
```

`k = 16,777,216` (`2**24`) completes in ~0.07s; `k = 16,777,217` hangs indefinitely (I killed it after several minutes while investigating).

Sweeping `m` at the same `k = 16,777,217`, the pattern is not monotonic:

| `m` | result |
|---|---|
| 2 | hangs |
| 3 | completes, ~3.57s (vs ~0.07s below the `k` threshold - ~50x slower, but returns) |
| 4 | completes, ~3.58s (consistent across repeats, scales to 5.31s at `k=25_000_000`) |
| 5 | completes, ~3.59s |
| 6 | hangs |
| 7 | hangs |
| 8 | hangs |

So `m=3,4,5` recover (slowly) while `m=2` and `m=6,7,8` hang outright - this looks like oneDNN is selecting different GEMM kernel/tiling implementations depending on `m`, and only some of those code paths hit the bad kernel-generation path.

`float32` does not reproduce at the analogous byte-size boundary: `k = 2**25` and `k = 2**25 + 1` (128 MiB and 128 MiB + 4 bytes, same as the float64 boundary in bytes) both complete in well under 0.1s. So whatever the actual overflowing value is, it isn't simply "total bytes moved" scaled the same way across dtypes.

Verbose log right before the hang (`m=2`, `k=16777217`, float64):

```
onednn_verbose,v1,info,oneDNN v3.12.0 (commit 80afa71049cd69a3df32adcccb623b12cd7baa22)
onednn_verbose,v1,info,gpu,runtime:SYCL
onednn_verbose,v1,info,gpu,engine,0,backend:Level Zero,name:Intel(R) Arc(TM) B390 GPU,driver_version:1.15.38646,binary_kernels:enabled
onednn_verbose,v1,primitive,error,gpu,jit::gemm,Invalid immediate value,src/gpu/intel/gemm/jit/gen_kernel.cpp:1035
onednn_verbose,v1,primitive,error,gpu,jit::gemm,Invalid immediate value,src/gpu/intel/gemm/jit/gen_kernel.cpp:1035
... (repeats, no forward progress)
```

**Nightly:** tried `torch==2.14.0.dev20260803+xpu` - it no longer hangs at any of the `m` values above (all of `m=2..8` now complete, consistently ~3.58-3.63s, vs ~0.04-0.07s below the `k` threshold). So the infinite-retry part seems to already be fixed there, but the underlying ~50-80x slowdown once `k` crosses `2**24` is still present across the board - looks like whatever changed stopped it from retrying forever, not the actual cost of whatever's failing/falling back once the immediate value doesn't fit.

**How I got here:** this showed up as a real hang in `sklearn.linear_model.Ridge` with `array_api_dispatch=True` (`device="xpu"`, torch backend, see array [API support doc](https://scikit-learn.org/dev/modules/array_api.html)) on `make_regression(n_samples=25_000_000, n_features=2)`. `Ridge`'s SVD solver computes `U.T @ y` after the SVD, where `U.T` ends up `(2, 25_000_000)`.
Since this was inside an unattended benchmark harness with a multi-minute timeout, letting it run long enough actually took down the GPU's compute engine, not just the one op:

```
kernel: xe 0000:00:02.0: [drm] Tile0: GT0: Engine reset: engine_class=ccs, logical_mask: 0x1, guc_id=28, state=0x3
kernel: xe 0000:00:02.0: [drm] Tile0: GT0: Timedout job: seqno=..., lrc_seqno=..., guc_id=28, flags=0x20 in python [pid]
kernel: xe 0000:00:02.0: [drm] exec queue reset detected
```

### Versions

```
PyTorch version: 2.13.0+xpu
Is debug build: False
CUDA used to build PyTorch: None

OS: Ubuntu 24.04.4 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Python version: 3.12.13 (64-bit runtime)
Python platform: Linux-7.0.0-28-generic-x86_64-with-glibc2.39
Is XPU available: True
XPU used to build PyTorch: 20260000
Intel GPU driver version:
* libze1: 1.28.6-1~24.04~ppa1
* intel-opencl-icd: 26.22.38646.6-1~24.04~ppa1
Intel GPU models detected:
* [0] _XpuDeviceProperties(name='Intel(R) Arc(TM) B390 GPU', platform_name='Intel(R) oneAPI Unified Runtime over Level-Zero V2', type='gpu', device_id=0xB080, driver_version='1.15.38646+6', total_memory=29304MB, is_integrated_gpu=1)

CPU: Intel(R) Core(TM) Ultra X7 358H, 16 cores
```

(`is_integrated_gpu=1` above - this is a Panther Lake iGPU, `Arc B390` is just the device name string oneDNN/torch report for it, not a discrete card.)

**Comment by LuFinch** [comment]
@cakedev0 Thanks for your detail issue description, I can reproduce this issue. 

I open a JIRA MFDNN-15428 and will update once OneDNN team has response..

**Comment by vpirogov** [comment]
Fixed by https://github.com/uxlfoundation/oneDNN/pull/5766. I can tag v3.12.4 with this fix if necessary.

**Comment by LuFinch** [comment]
@vpirogov Thanks for your quick action. I verified that the fix could help eliminate the warning and performance regression when dim k crosses 2**24. However, this Pytorch2.14 release schedule is tight so that we don't have enough time to do an round acceptance test for v3.12.4 upgrade.

For the hang issue, it was already solved in Pytorch2.13.
For the remained performance issue, we would like to solve it in Pytorch2.15 with OneDNN v3.13 upgrade.

**Comment by chuanqi129** [comment]
Thanks @cakedev0 for the issue report, the performance regression fix will target on PyTorch 2.15
